### PR TITLE
feat: IOx metrics crate

### PR DESCRIPTION
**Pull request by e-dard**
## Rationale

In my experience once many teams start running many different services within the same environment (cluster, platform, infrastructure etc) unless some reasonably consistent procedure is put in place on instrumenting services one ends up with a big homogeneous mess. Both within a single application and across many applications.

So, this PR introduces a new crate that I would love to see being the single API used to add telemetry that we want to expose via IOx's Prometheus endpoint. If we move away from Prometheus it's a single place to change the plumbing. If we want to add a new type of metric or a new way of labeling/defining things, it's one place to do it.

The other thing I have seen is that it can be tricky to test metrics when they're (1) global; (2) not consistently created/categorised/initialised etc. Testing metrics is really important.

So the approach taken in this PR intends to make it easier to actually test your metrics in IOx.

## Approach

The TLDR is:

 - single crate (`metrics` - don't mind what the name is).
 - the IOx application will initialise a single `MetricRegistry` very high up in `main`. This registry will be pushed down to all sub-systems. It is not a static/global.
 - We define the concept of a _domain_. A domain just means that your metrics on that domain will have a common prefix. This is useful when your metrics inevitably get mixed around with others in a sink somewhere. But it's also just easier for people to get context on metrics. Examples of domains might be: `http`, `catalog`, `wb`, `mub`, `rb` etc.
 - You have to register domains on the `MetricRegistry`, which means your code needs to be provided with an instance of it. We are trading of simplicity of a static for a bit of plumbing, but we get better isolation and testability.
 - With a registered domain you can then register metrics. When you register a metric you get to provide things like a name, unit, description and what type of metric you want.
 - In the PR there are two types of metrics: `REDMetic` and `Counter`. I plan on adding gauges too but not got there yet.
 - A `REDMetric` should be used for anything you might need to have measure against an SLO, e.g., the availabilty of the write path. A `REDMetric` is an abstraction over a counter for requests and whether they succeeded or failed, and a histogram tracking request latencies. You get a single API and it handles all the implementation for you.
 - A `Counter` is a monotonically increasing counter metric. You can give it a name, unit etc. It's useful for tracking things like "how many bytes have we ingested" or "how much data have we moved to the read buffer".
 - Metrics can have _default labels_ (labels associated with all observations) or you can provide some labels on an observation by observation basis.

Finally, I created a `TestMetricRegistry`, which just wraps a `MetricRegistry`. The idea with this is that is provides you with an API for asserting conditions against metrics, which should make testing a bit simpler. But you can also just test against the emitted prometheus exposition format if you wanted to.

## What's in the PR?

- I have created an initial version of the above approach.
- I have converted all existing metrics (mainly tracking line protocol on the write path) and converted them to using the new way.
- I have implemented more metrics on IOx's HTTP API covering a bunch of other endpoints: query, snapshotting, write_buffer_meta, metrics etc.
- I've removed the previous metrics implementation.

## What's not in this PR?

This PR is already big, I want to pause and get feedback on the general idea. Perhaps merge the PR and then move forward with improvements afterwards. With that in mind, here are a list of things that could be improved.

- [ ] I didn't come up with a good story for handling early returns from functions using `?`. So in this PR a lot of error conditions will result in missing metrics. I can imagine either macros solving this or maybe another way. I'm sure people have good ideas.
- [ ] The code in the `metrics` crate could easily be more DRY using macros. I didn't bother as it's easier and quicker to copy/paste. I can delete code with macros later 😉.
- [ ] There are no gauge metrics yet because I couldn't figure out the right OT metric to use (that also accepted labels). Once that's figured out I plan to add that abstraction too.
- [ ] I suspect we can merge the `observability_deps` into this crate. I'll check what other folks think.

### CODE REVIEW

File: metrics/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+///
+/// Each sub-system should register a _domain_, which will result in all metrics
+/// for that domain having a common name prefix. Domains allow callers to
+/// register individual metrics.
+///
+/// To test your metrics in isolation simply provide a new `TestMetricRegistry`
+/// during test setup. As well as ensuring isolation of your metrics it also
+/// provides handy builder to easily assert conditions on your metrics.
+#[derive(Debug)]
+pub struct MetricRegistry {
```
Comment (e-dard) [inline metrics/src/lib.rs:43]: IOx will make one of these in the application and it gets pushed down to various sub-systems (domains).

In tests each sub-system can initialise a new one (or better yet use `TestMetricRegistry`) to get metric test isolation.

File: metrics/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+    /// # HELP a_value_recorder Records values
+    /// # TYPE a_value_recorder histogram
+    /// a_value_recorder_bucket{key="value",le="0.5"} 0
+    /// a_value_recorder_bucket{key="value",le="0.9"} 0
+    /// a_value_recorder_bucket{key="value",le="0.99"} 0
+    /// a_value_recorder_bucket{key="value",le="+Inf"} 1
+    /// a_value_recorder_sum{key="value"} 100
+    /// a_value_recorder_count{key="value"} 1
+    ///
+    pub fn metrics_as_text(&self) -> Vec<u8> {
```
Comment (e-dard) [inline metrics/src/lib.rs:74]: This provides the metrics in the prometheus exposition format.

File: metrics/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    ///
+    /// `mydomain.requests.total` (tracks total and failed requests)
+    /// `mydomain.requests.duration.seconds` (tracks a distribution of
+    /// latencies)
+    ///
+    /// If `name` is provided then the metric names become:
+    //
+    /// `mydomain.somename.requests.total`
+    /// `mydomain.somename.requests.duration.seconds`
+    pub fn register_red_metric(&self, subname: Option<&str>) -> metrics::RedMetric {
```
Comment (e-dard) [inline metrics/src/lib.rs:173]: This API effectively limits what kind of metrics you can have in IOx and how they're named. Consumers of the API provide metric-specific details around naming and description etc and the API handles initialising them in a consistent way w.r.t our metrics backend (prometheus) and also naming conventions.

File: metrics/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::time::Duration;
+
+    #[test]
+    fn red_metric() {
```
Comment (e-dard) [inline metrics/src/lib.rs:248]: This is probable the best place to understand how the whole crate works.

File: metrics/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+        let reg = TestMetricRegistry::default();
+        let domain = reg.registry().register_domain("http");
+
+        // create a RED metrics
+        let metric = domain.register_red_metric(None);
+
+        // Get an observation to start measuring something.
+        let ob = metric.observation();
```
Comment (e-dard) [inline metrics/src/lib.rs:256]: When this is called it starts timing the rest of the function execution until you call a finishing method on `ob`.

File: metrics/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+        // Usually caller would call `ob.ok()`.
+        ob.observe(metrics::RedRequestStatus::Ok, duration, &[]);
+
+        reg.has_metric_family("http_requests_total")
+            .with_labels(&[("status", "ok")])
+            .counter()
+            .eq(1.0)
+            .unwrap();
+
+        reg.has_metric_family("http_request_duration_seconds")
```
Comment (e-dard) [inline metrics/src/lib.rs:273]: I'm interesting in what people think about this builder style of test asseting. I sort of like it but I think it could be a bit smarter and less verbose. I know @domodwyer has some good ideas here...

File: metrics/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+            metrics::RedRequestStatus::Error,
+            Duration::from_millis(350),
+            &[],
+        );
+
+        // You are always welcome to assert against the exposition format
+        // directly.
+        assert_eq!(
+            String::from_utf8(reg.registry().metrics_as_text()).unwrap(),
+            vec![
```
Comment (e-dard) [inline metrics/src/lib.rs:304]: You can always just assert against the prom text if you like.

File: metrics/src/metrics.rs [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+///
+/// The RED methodology stipulates you should track three key measures:
+///
+///  - Request Rate: (total number of requests / second);
+///  - Error Rate: (total number of failed requests / second);
+///  - Duration: (latency distributions for the various requests)
+///
+/// Using a `REDMetric` makes following this methodology easy because it handles
+/// updating the three components for you.
+pub struct RedMetric {
```
Comment (e-dard) [inline metrics/src/metrics.rs:62]: This is an abstraction over two metics: (1) a counter that tracks successful interactions; (2) a histogram that tracks latencies. With this abstraction you can follow the [RED methodology](https://www.weave.works/blog/the-red-method-key-metrics-for-microservices-architecture/).

File: metrics/src/metrics.rs [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    /// Timing of observation is handled automatically.
+    pub fn ok_with_labels(&self, labels: &[KeyValue]) {
+        let duration = self.start.elapsed();
+        self.observe(RedRequestStatus::Ok, duration, labels);
+    }
+
+    /// Record that the observation was not successful but was still valid.
+    /// `client_error` is the right thing to choose when the request failed perhaps
+    /// due to client error. Timing of observation is handled automatically.
+    pub fn client_error(&self) {
```
Comment (e-dard) [inline metrics/src/metrics.rs:186]: One thing that might seems a bit unusual or weird is the notion that there is a difference between errors caused by the service/system being instrumented and errors caused by external input. An example of the latter would be things that result in `4XX` errors. The reason why I think it's good to differentiate between the two is that it can be useful to observe where customers/downstream services are not interacting with your service/system correctly.

File: server/src/lib.rs [hunk 10]
```diff
@@ -237,6 +241,41 @@ impl ServerConfig {
     }
 }
 
+// A collection of metrics used to instrument the Server.
+#[derive(Debug)]
+pub struct ServerMetrics {
```
Comment (e-dard) [inline server/src/lib.rs:246]: These are the old metrics but also covering more surface area (e.g., all API requests). I will update the IOx dashboards accordingly when this gets merged.

File: server/src/lib.rs [hunk 11]
```diff
@@ -842,15 +899,19 @@ mod tests {
...
 
     use super::*;
     use std::sync::atomic::{AtomicBool, Ordering};
 
     fn config() -> ServerConfig {
-        ServerConfig::new(Arc::new(ObjectStore::new_in_memory(InMemory::new())))
-            .with_num_worker_threads(1)
+        ServerConfig::new(
+            Arc::new(ObjectStore::new_in_memory(InMemory::new())),
+            Arc::new(MetricRegistry::new()), // new registry ensures test isolation of metrics
```
Comment (e-dard) [inline server/src/lib.rs:912]: Effectively we use a different registry for each test, so we can get metric isolattion :tad

File: src/influxdb_ioxd/http.rs [hunk 12]
```diff
@@ -421,8 +420,15 @@ async fn write<M>(req: Request<Body>) -> Result<Response<Body>, ApplicationError
 where
     M: ConnectionManager + Send + Sync + Debug + 'static,
 {
+    let path = req.uri().path().to_string();
     let server = Arc::clone(&req.data::<Arc<AppServer<M>>>().expect("server state"));
 
+    // TODO(edd): figure out best way of catching all errors in this observation.
+    let obs = server.metrics.http_requests.observation(); // instrument request
```
Comment (e-dard) [inline src/influxdb_ioxd/http.rs:427]: Need a smarter person than me to come up with a good way to observe early returns :-)

File: src/influxdb_ioxd/http.rs [hunk 13]
```diff
@@ -464,14 +475,13 @@ where
...
+    );
 
-    IOXD_METRICS
-        .lp_lines_success
-        .add(lines.len() as u64, &metric_kv);
-    IOXD_METRICS
-        .lp_bytes_success
-        .add(body.len() as u64, &metric_kv);
-
+    obs.ok_with_labels(&metric_kv); // request completed successfully
```
Comment (e-dard) [inline src/influxdb_ioxd/http.rs:484]: This will time the request and incremement the counter.

File: metrics/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+    fn build_metric_prefix(&self, metric_name: &str, subname: Option<&str>) -> String {
+        match subname {
+            Some(subname) => format!("{}.{}.{}", self.name, subname, metric_name),
+            None => format!("{}.{}", self.name, metric_name),
+        }
+    }
+
+    /// Registers a new metric following the RED methodology.
+    ///
+    /// By default, two distinct metrics will be created. One will be a counter
```
Comment (jacobmarble) [inline metrics/src/lib.rs:157]: "By default" should be removed. There are two distinct metrics by design, and the user can't change that.

File: metrics/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+/// for that domain having a common name prefix. Domains allow callers to
+/// register individual metrics.
+///
+/// To test your metrics in isolation simply provide a new `TestMetricRegistry`
+/// during test setup. As well as ensuring isolation of your metrics it also
+/// provides handy builder to easily assert conditions on your metrics.
+#[derive(Debug)]
+pub struct MetricRegistry {
+    provider: RegistryMeterProvider,
+    exporter: PrometheusExporter,
```
Comment (jacobmarble) [inline metrics/src/lib.rs:45]: The exporter should be pluggable and configurable, but I don't think now is the time to do that, as long as we don't tie ourselves only to Prometheus.

File: metrics/src/metrics.rs [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+        let record = move |status: RedRequestStatus, duration: Duration, labels: &[KeyValue]| {
+            let labels = match labels.is_empty() {
+                // If there are no labels specified just borrow defaults
+                true => Cow::Borrowed(&self.default_labels),
+                false => {
+                    // Otherwise merge the provided labels and the defaults.
+                    // Note: provided labels need to go last so that they overwrite
+                    // any default labels.
+                    //
+                    // PERF(edd): this seems expensive to me.
```
Comment (jacobmarble) [inline metrics/src/metrics.rs:104]: Interesting that the OTel SDK chose "default" instead of "const".

File: metrics/src/metrics.rs [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+
+pub use observability_deps::opentelemetry::KeyValue;
+
+const RED_REQUEST_STATUS_LABEL: &str = "status";
+
+/// Possible types of RED metric observation status.
+///
+/// Ok      - an observed request was successful.
+/// OkError - an observed request was unsuccessful but it was not the fault of
+///           the observed service.
```
Comment (jacobmarble) [inline metrics/src/metrics.rs:19]: Is this a convention you have seen somewhere else? Without this comment, it isn't obvious.

The distinction is similar to 4XX vs 5XX, is that right? Sometimes those are bucketed "invalid request" and "retryable error".

File: metrics/src/lib.rs (hunk 15 above)
Comment (e-dard) [reply metrics/src/lib.rs:45]: Filed in https://github.com/influxdata/influxdb_iox/issues/1279

File: metrics/src/metrics.rs (hunk 17 above)
Comment (e-dard) [reply metrics/src/metrics.rs:19]: I meant to update this to `ClientError` before I pushed the PR. Have done now.

File: src/influxdb_ioxd/http.rs (hunk 12 above)
Comment (domodwyer) [reply src/influxdb_ioxd/http.rs:427]: There's a couple ways I can think of - the nicest dev experience would be to provide something like the the tracing [`#[instrument]`](https://docs.rs/tracing/0.1.25/tracing/attr.instrument.html) macro but it's definitely not trivial, and has some async gotchas.

The easiest way I can think of is to provide an extension trait to the `Result` type with something [like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d7861cb988f40a9f2ee532488ae9e93b) and then decorate the code paths with it where we want the error counter to be incremented when an early return is taken. We could have both an `observe()` and `observe_client()` to differentiate between the two error classes, assuming an error maps cleanly to one of them.

At first I naively thought a `Drop` impl that defaults to an error might be useful, but it'll be horribly brittle and easy to naff up.

File: metrics/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,428 @@
...
+    /// Registers a new counter metric.
+    ///
+    /// `name` should be a noun that describes the thing being counted, e.g.,
+    /// lines, payloads, accesses, compactions etc.
+    ///
+    /// `unit` is optional and will appear in the metric name before a final
+    /// `total` suffix. Consider reviewing
+    /// https://prometheus.io/docs/practices/naming/#base-units for appropriate
+    /// units.
+    pub fn register_counter_metric(
```
Comment (domodwyer) [inline metrics/src/lib.rs:207]: It might be nicer (and more future proof) to avoid having `register_counter_metric()`, `register_counter_metric_with_labels()` and other combinations of options, particularly that they also duplicate optional fields (`unit`).

An alternative strategy is to use the builder pattern to expose an interface like:

```rust
    // A simple counter
    let counter = reg
        .register_counter_metric("name", "A description of my metric")
        .build();

    // With some units
    let counter = reg
        .register_counter_metric("name", "A description of my metric")
        .with_unit("bytes/s")
        .build();
    
    // With some units and default labels
    let counter = reg
        .register_counter_metric("name", "A description of my metric")
        .with_unit("bytes/s")
        .with_default_labels(&[....])
        .build();
```

File: metrics/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,428 @@
...
+
+        let provider = exporter.provider().unwrap();
+
+        Self { provider, exporter }
+    }
+}
+
+/// A `Domain` provides a namespace that describes a collection of related
+/// metrics. Within the domain all metrics' names will be prefixed by the name
+/// of the domain; for example "http", "read_buffer", "wal".
```
Comment (carols10cents) [inline metrics/src/lib.rs:131]: s/wal/write_buffer or wb/g please :)

File: metrics/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,428 @@
...
+        description: impl Into<String>,
+    ) -> metrics::Counter {
+        self.register_counter_metric_with_labels(name, unit, description, vec![])
+    }
+
+    /// Registers a new counter metric with default labels.
+    pub fn register_counter_metric_with_labels(
+        &self,
+        name: &str,
+        unit: Option<String>,
```
Comment (carols10cents) [inline metrics/src/lib.rs:220]: The type for `unit` in these methods can probably be `Option<&str>` as you did with `name` in the other methods

File: metrics/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,428 @@
...
+        let counter = self
+            .meter
+            .u64_counter(format!(
+                "{}{}.total",
+                self.build_metric_prefix(name, None),
+                match unit {
+                    Some(unit) => format!(".{}", unit),
+                    None => "".to_string(),
+                }
+            ))
```
Comment (carols10cents) [inline metrics/src/lib.rs:233]: I think the way you did this in `build_metric_prefix` is better, how you put the `format!` within the `match` rather than having 2 `format!`s

File: metrics/src/lib.rs (hunk 6 above)
Comment (carols10cents) [reply metrics/src/lib.rs:272]: I like builders 🤷🏻‍♀️

File: metrics/src/metrics.rs [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+impl RedMetric {
+    pub(crate) fn new(
+        requests: OTCounter<u64>,
+        duration: OTHistorgram<f64>,
+        mut default_labels: Vec<KeyValue>,
+    ) -> Self {
+        // TODO(edd): decide what to do if `labels` contains
+        // RED_REQUEST_STATUS_LABEL.
```
Comment (domodwyer) [inline metrics/src/metrics.rs:77]: This is unfortunate, I imagine "status" is not unlikely to be used! Perhaps we can simply prefix it as `_status`  or use `red_status` or something to make it less likely to happen?

We can't panic here because it's possible the labels wind up coming from some sort of (sanitised) user input like "DB name" or something, which could legitimately be called 'status'.

File: metrics/src/metrics.rs [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+            };
+
+            match status {
+                RedRequestStatus::Ok => {
+                    self.requests.add(1, &labels);
+                    self.duration.record(duration.as_secs_f64(), &labels);
+                }
+                RedRequestStatus::ClientError => {
+                    let mut labels = labels.into_owned();
+                    labels[0] = KeyValue::new(
```
Comment (domodwyer) [inline metrics/src/metrics.rs:120]: It really feels like `Vec` isn't the right container here... But I guess we're restricted by what the underlying library uses?

File: metrics/src/metrics.rs [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+            requests,
+            duration,
+            default_labels,
+        }
+    }
+
+    /// Returns a new observation that will handle timing and recording an
+    /// observation the metric is tracking.
+    pub fn observation(
+        &'_ self,
```
Comment (carols10cents) [inline metrics/src/metrics.rs:94]: I don't think this `'_` on `self` is needed, it compiles for me without it...

```suggestion
        &self,
```

File: metrics/src/metrics.rs [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+                false => {
+                    // Otherwise merge the provided labels and the defaults.
+                    // Note: provided labels need to go last so that they overwrite
+                    // any default labels.
+                    //
+                    // PERF(edd): this seems expensive to me.
+                    let mut new_labels: Vec<KeyValue> = self.default_labels.clone();
+                    new_labels.extend(labels.iter().cloned());
+                    Cow::Owned(new_labels)
+                }
```
Comment (carols10cents) [inline metrics/src/metrics.rs:110]: Matching on one bool is a bit clearer as an if/else, IMO :) And you can use `extend_from_slice` to make this a bit shorter, I don't think it makes it more performant tho. It might not be bad because it looks like `KeyValue` uses `Cow` under the hood... profiling needed! :)

```suggestion
            let labels = if labels.is_empty() {
                // If there are no labels specified just borrow defaults
                Cow::Borrowed(&self.default_labels)
            } else {
                // Otherwise merge the provided labels and the defaults.
                // Note: provided labels need to go last so that they overwrite
                // any default labels.
                //
                // PERF(edd): this seems expensive to me.
                let mut new_labels: Vec<KeyValue> = self.default_labels.clone();
                new_labels.extend_from_slice(labels);
                Cow::Owned(new_labels)
```

File: metrics/src/metrics.rs [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+
+impl Display for RedRequestStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Ok => write!(f, "ok"),
+            Self::ClientError => write!(f, "client_error"),
+            Self::Error => write!(f, "error"),
+        }
+    }
+}
```
Comment (carols10cents) [inline metrics/src/metrics.rs:50]: Because these variants have string literal values, I think we can improve the performance a bit below by getting rid of some of the `.to_string`s: I'd add this:

```
impl RedRequestStatus {
    fn as_str(&self) -> &'static str {
        match self {
            Self::Ok => "ok",
            Self::ClientError => "client_error",
            Self::Error => "error",
        }
    }
}
```

Then this `Display` impl can just call `write!(f, self.as_str())`. Everywhere below that you call "to_string" on one of these variants can then be `as_str` instead, and then the `Cow`s will be `Cow<&'static str>` rather than `Cow<String>`.

Hard to say without profiling though :)

File: metrics/src/metrics.rs [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+        };
+
+        RedObservation::new(record)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RedObservation<T>
+where
+    T: Fn(RedRequestStatus, Duration, &[KeyValue]),
```
Comment (domodwyer) [inline metrics/src/metrics.rs:148]: This a slightly unusual way to allow a "pluggable" collector - I'd probably make a trait that defines what methods a collector needs to provide, and then be generic over that - [here's an example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=27ff960d46677f3c1b87a943da181013):

```rust
// Instead of invoking a Fn, make the RedObservation be generic over an 
// implementation of something that records observations.
//
// Describe the behaviour of something that records observations in this trait.
pub trait RedCollector {
    // A method to invoke when the RedObservation is finished to record the results.
    fn record(&self, status: RedRequestStatus, d: Duration, kv: &[KeyValue]);
}
```

And then make the `RedObservation` generic over it:

```rust
pub struct RedObservation<T>
where
    T: RedCollector,
{
    metric: T,
    // instant, etc goes here as normal
}
```

You can then implement `RedCollector` for _a reference to_ `RedMetric` - this lets you give out `RedObservation` instances that trigger a callback on the `RedMetric` itself without giving up ownership / using an arc:

```rust
impl RedCollector for &RedMetric {
    fn record(&self, status: RedRequestStatus, d: Duration, kv: &[KeyValue]) {
        // We have access to the RedMetric instance as &self
    }
}
```

**Review by carols10cents — APPROVED** [review]
A few thoughts, and I agree with Dom that taking a look at what `#[instrument]` does and seeing if that's applicable would be good for not needing to remember to call methods at the end of a function to collect the right metrics. 

I'd selfishly like this to be merged in because I get a test failure locally having to do with metrics on `main`, but I don't on this branch ;)

**Review by domodwyer — APPROVED** [review]
I love the approach - enforcing consistency across the platform by using a carefully designed API to make it easy to "do the right thing" is great in my book :ok_hand:

I particularly love the work put in to make using metrics _testable_ - there's no point having metrics if they're wrong, so testing them is as important as having them! The easier it is, the more likely it'll be tested and it'll get better coverage as a result. It also provides an easy way for contributors to validate correct code paths are being taken when writing tests if they're not super experienced with rust or testing & asserting with mocks, etc - invaluable for an open source project!

My comments are all about ergonomics for end users, or stylistic rusty things - nothing to stop this going forward!

File: metrics/src/tests.rs [hunk 28]
```diff
@@ -0,0 +1,458 @@
...
+    /// family. Use the family name, e.g., `http_request_duration_seconds` to
+    /// get access to the individual metrics via the `histogram` method on the
+    /// returned `AssertionBuilder`.
+    pub fn try_has_metric_family<'a>(&'a self, name: &str) -> Result<AssertionBuilder<'a>, Error> {
+        let metric_families = self.registry.exporter.registry().gather();
+        let family = metric_families
+            .into_iter()
+            .find(|fam| fam.get_name() == name)
+            .context(MetricFamilyNotFoundError {
+                name: name.to_owned(),
```
Comment (shepmaster) [inline metrics/src/tests.rs:71]: Function calls eagerly evaluate their arguments, which means that this `to_owned` call will happen unconditionally. However, `context` takes its argument and calls `into` on each of the fields, which allows certain things to behave lazier.

TL;DR:

```suggestion
                name,
```

File: metrics/src/tests.rs [hunk 29]
```diff
@@ -0,0 +1,458 @@
...
+                    name: self.family.get_name().to_owned(),
+                    labels: self.labels.clone(),
+                    metrics: self.registry.metrics_as_str(),
+                }
+                .fail(),
+                family_name: "".to_string(),
+                metric_dump: "".to_string(),
+            };
+        }
+        let metric = metric.unwrap();
```
Comment (shepmaster) [inline metrics/src/tests.rs:203]: It's an antipattern to check if something is `None` and then unwrap it if not. Instead, something like:

```suggestion
        let metric = match metric {
            Some(metric) => metric,
            None => {
            return Histogram {
                c: NoMatchingLabelsError {
                    name: self.family.get_name(),
                    labels: self.labels.clone(), // Maybe `labels: &self.labels`
                    metrics: self.registry.metrics_as_str(),
                }
                .fail(),
                family_name: "".to_string(),
                metric_dump: "".to_string(),
            };
        };
```

File: metrics/src/tests.rs [hunk 30]
```diff
@@ -0,0 +1,458 @@
...
+        let c = self.c?; // return previous errors
+
+        if v != c.get_value() {
+            return FailedMetricAssertionError {
+                name: self.family_name,
+                msg: format!("{:?} == {:?} failed", c.get_value(), v),
+                metrics: self.metric_dump,
+            }
+            .fail();
+        }
```
Comment (shepmaster) [inline metrics/src/tests.rs:247]: This (and similar future cases in this file) could probably use the snafu `ensure!` macro. Something like:

```suggestion
        ensure!(v == c.get_value(), FailedMetricAssertionError {
            name: self.family_name,
            msg: format!("{:?} == {:?} failed", c.get_value(), v),
            metrics: self.metric_dump,
        });            
```

File: metrics/src/metrics.rs (hunk 23 above)
Comment (e-dard) [reply metrics/src/metrics.rs:120]: yeah a lot of this feels like a map is more appropriate. The prom APIs all take a slice of labels. I suppose I could add a map at the `metrics` crate level and convert to a vec only inside this crate.

File: metrics/src/lib.rs [hunk 31]
```diff
@@ -0,0 +1,428 @@
...
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::time::Duration;
+
+    #[test]
+    fn red_metric() {
+        let reg = TestMetricRegistry::default();
+        let domain = reg.registry().register_domain("http");
```
Comment (jacobmarble) [inline metrics/src/lib.rs:249]: Could domains be recursive? `let subdomain = domain.register_domain("subdomain")`

File: metrics/src/lib.rs [hunk 32]
```diff
@@ -0,0 +1,428 @@
...
+        let metric = domain.register_red_metric(None);
+
+        // Get an observation to start measuring something.
+        let ob = metric.observation();
+
+        // Usually a caller would use `ob.ok_with_labels(labels)`;
+        ob.observe(
+            metrics::RedRequestStatus::Ok,
+            Duration::from_millis(100),
+            &[KeyValue::new("account", "abc123")],
```
[the remark this answers is not in the document]
Comment (jacobmarble) [reply metrics/src/lib.rs:350]: FYI the Otel data model is about to drop `KeyValue` (string:string) in favor of `Attribute` (string:typed-value).

[the remark this answers is not in the document]
Comment (jacobmarble) [reply metrics/src/lib.rs:350]: Otel-proto has now dropped string:string labels in favor of string:typed attributes.
https://github.com/open-telemetry/opentelemetry-proto/pull/283

[the remark this answers is not in the document]
Comment (jacobmarble) [reply metrics/src/lib.rs:350]: ...and since we're just supporting Prometheus for now, it's probably best to keep the string:string abstraction until that changes. Later, string:typed is a great way to show how cool IOx is for storing metrics.

File: src/influxdb_ioxd/http.rs (hunk 12 above)
Comment (shepmaster) [reply src/influxdb_ioxd/http.rs:427]: > provide something like the the tracing [`#[instrument]`](https://docs.rs/tracing/0.1.25/tracing/attr.instrument.html) macro

Speaking from a place of true ignorance, would it make sense to create our own tracing subscriber? Then users could use the `#[instrument]` macro and shunt the data into us.

Comment (domodwyer) [reply src/influxdb_ioxd/http.rs:427]: I really wanted this to work too - I think the biggest stopper for this idea is that there's no obvious way to tie an `#[instrument]`'ed call to a particular metric. As an example, an API like this would be great:

```rust
#[instrument(err, metric="http.requests")]
fn thinger() -> Result<(), Error> {
    todo!()
}
```

And in order to capture the "user error vs our error" concept Edd is trying to collect data on, it'd likely have to be expanded further:

```rust
#[instrument(err, metric="http.requests", err_type=client)]
fn thinger() -> Result<(), Error> {
    todo!()
}
```

From what I could see, there was no way for a subscriber to see what aribitrary attributes are set in the macro invocation (which may be a nice idea for allowing tracing subscribers to provide expanded functionality, such as differentiating between classes of spans).

I _think_ it might be possible to hack around this by overloading the `target` or adding a special `field` to the span and filtering on the [metadata](https://docs.rs/tracing-core/0.1.17/tracing_core/metadata/struct.Metadata.html) the subscriber is given for the [new_span()](https://docs.rs/tracing/0.1.25/tracing/trait.Subscriber.html#tymethod.new_span) call:

```rust
#[instrument(err, fields(metric.name = "http.requests")]
fn thinger() -> Result<(), Error> {
    todo!()
}
```

But I didn't look into if/how we could handle the errors with it as it started to feel a bit magic... Perhaps there's a nice way to wrap it up?

File: metrics/src/lib.rs (hunk 18 above)
Comment (e-dard) [reply metrics/src/lib.rs:207]: nice. We should do something like this

File: metrics/src/metrics.rs (hunk 22 above)
Comment (e-dard) [reply metrics/src/metrics.rs:83]: yeah, need to think about that.

File: metrics/src/tests.rs (hunk 29 above)
Comment (e-dard) [reply metrics/src/tests.rs:203]: not sure what I was doing there. Thanks

File: metrics/src/lib.rs [hunk 33]
```diff
@@ -0,0 +1,428 @@
...
+    pub fn metrics_as_text(&self) -> Vec<u8> {
+        let metric_families = self.exporter.registry().gather();
+        let mut result = Vec::new();
+        TextEncoder::new()
+            .encode(&metric_families, &mut result)
+            .unwrap();
+        result
+    }
+
+    pub(crate) fn metrics_as_str(&self) -> String {
```
[the remark this answers is not in the document]
Comment (alamb) [reply metrics/src/lib.rs:85]: `metrics_as_string` perhaps

File: src/influxdb_ioxd/http.rs [hunk 34]
```diff
@@ -421,8 +420,15 @@ async fn write<M>(req: Request<Body>) -> Result<Response<Body>, ApplicationError
...
     M: ConnectionManager + Send + Sync + Debug + 'static,
 {
+    let path = req.uri().path().to_string();
     let server = Arc::clone(&req.data::<Arc<AppServer<M>>>().expect("server state"));
 
+    // TODO(edd): figure out best way of catching all errors in this observation.
+    let obs = server.metrics.http_requests.observation(); // instrument request
+
+    // TODO - metrics. Implement a macro/something that will catch all the
+    // early returns.
```
Comment (alamb) [inline src/influxdb_ioxd/http.rs:430]: I wonder if we could implement a RAII / guard that recorded an error on `drop` if no observations had been made

File: metrics/src/lib.rs [hunk 35]
```diff
@@ -0,0 +1,426 @@
...
+    pub fn register_domain(&self, name: &'static str) -> Domain {
+        let meter = self.provider.meter(name, None);
+        Domain::new(name, meter)
+    }
+}
+
+impl Default for MetricRegistry {
+    fn default() -> Self {
+        let registry = Registry::new();
+        let default_histogram_boundaries = vec![0.5, 0.9, 0.99];
```
Comment (mkmik) [inline metrics/src/lib.rs:105]: Uh I just noticed this; 

The histogram boundaries should define buckets in the domain of the value being measured.

For example, if you're measuring duration it should be some reasonable durations for the request.

This _seems_ as it's defining common quantiles, (p50, p90 and p99).

We won't get useful histogram quantiles out of this kind of bucketing, since most requests will fall outside of the range 500ms < d < 1s

Comment (alamb) [reply metrics/src/lib.rs:105]: Proposed fix in https://github.com/influxdata/influxdb_iox/pull/1322
